src/engine/strat_engine/cmd.rs

```diff
@@ -0,0 +1,155 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Handles invoking external binaries.
+// This module assumes that, for a given machine, there is only one place
+// where the desired executable might be installed. It expects the engine
+// to identify that place at its initialization by invoking verify_binaries(),
+// and to exit immediately if verify_binaries() return an error. If this
+// protocol is followed then when any command is executed the unique absolute
+// path of the binary for this machine will already have been identified.
+// However stratisd may run for a while and it is possible for the binary
+// to be caused to be uninstalled while stratisd is being run. Therefore,
+// the existence of the file is checked before the command is invoked, and
+// an explicit error is returned if the executable can not be found.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use uuid::Uuid;
+
+use stratis::{ErrorEnum, StratisError, StratisResult};
+
+/// Find the binary with the given name by looking in likely locations.
+/// Return None if no binary was found.
+fn find_binary(name: &str) -> Option<PathBuf> {
+    ["/usr/sbin", "/sbin", "/usr/bin", "/bin"]
+        .iter()
+        .map(|pre| [pre, name].iter().collect::<PathBuf>())
+        .find(|path| path.exists())
+}
+
+// These are the external binaries that stratisd relies on.
+// Any change in this list requires a corresponding change to BINARIES,
+// and vice-versa.
+const MKFS_XFS: &str = "mkfs.xfs";
+const THIN_CHECK: &str = "thin_check";
+const THIN_REPAIR: &str = "thin_repair";
+const XFS_ADMIN: &str = "xfs_admin";
+const XFS_GROWFS: &str = "xfs_growfs";
+
+lazy_static! {
+    static ref BINARIES: HashMap<String, Option<PathBuf>> = [
+        (MKFS_XFS.to_string(), find_binary(MKFS_XFS)),
```

Comment:
For required binaries that are only needed during test what are you suggesting we do?

Reply:
* Make a separate HashMap to store the names and absolute paths of these test-only binaries.
* Don't bother to write a ```verify_binaries``` method. We have a deliberate policy of unwrapping all over in tests, that would be consistent with this policy.
* Add a method ```get_test_executable``` which has fewer ```expects()```s and more ```unwraps()``` than ```get_executable``` consistent w/ the way tests are run. Use ```get_test_executable()``` when getting executables in tests.

Reply:
Also, as much of the code that I outlined above should be modified with the ```#cfg[test]``` annotation.